Test Github Actions - ansible-lint

wiktor2200 opened the pull request:
Test implementation of Github Action for ansible-lint.
https://github.com/ansible-community/ansible-lint-action

wiktor2200 commented:
I guess it needs some more permission, but you can see results of this workflow on my forked repo: https://github.com/wiktor2200/install_nextcloud/actions

staticdev commented:
@wiktor2200 I have a complete pipeline for tests using GA that I use for all my roles.. it is an easy copy-paste and has some other advantages since it also run yaml-lint, molecule and have versions updated by dependabot automatically: https://github.com/staticdev/ansible-role-signal/blob/main/.github/workflows/tests.yml

staticdev commented on .github/workflows/ansible-lint.yml:
```diff
@@ -0,0 +1,19 @@
+## Source: https://github.com/marketplace/actions/ansible-lint
```
If you rename your workflow `tests.yml` the badge to README.md can be added with:

```md
|Tests|

.. |Tests| image:: https://github.com/aalaesar/install_nextcloud/workflows/Tests/badge.svg
   :target: https://github.com/aalaesar/install_nextcloud/actions?workflow=Tests
   :alt: Tests
```

wiktor2200 replied:
I've added ansible-lint and yamllint as two separate workflows so I've added two badges to README.

staticdev commented on .github/workflows/ansible-lint.yml:
```diff
@@ -0,0 +1,19 @@
+
+    steps:
+      # Important: This sets up your GITHUB_WORKSPACE environment variable
+      - uses: actions/checkout@v2
```
Latest version of checkout is v3.

wiktor2200 replied:
Good catch, I just copied default config from ansible-lint workflow.

wiktor2200 replied:
Done

staticdev commented on .github/workflows/ansible-lint.yml:
```diff
@@ -0,0 +1,19 @@
+      # Important: This sets up your GITHUB_WORKSPACE environment variable
+      - uses: actions/checkout@v2
+
+      - name: Run ansible-lint
```
To have 5 star on ansible-galaxy it is also needed to run yamllint together with ansible-lint. One can do it with:

```md
      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"

      - name: Install test dependencies
        run: pip install --constraint=.github/workflows/constraints.txt ansible yamllint ansible-lint

      - name: Lint code
        run: |
          yamllint .
          ansible-lint
```

Note that versions of yamllint and ansible-lint are pinned so a constants.txt file in the same folder with:

```
pip==22.1.2
ansible==6.0.0
yamllint==1.26.3
ansible-lint==6.3.0
```

staticdev reviewed (changes requested):
I have sent some extra suggestions.. feel free to pick whatever you like.

wiktor2200 commented:
@staticdev Thanks for your insights. :)
I've added two separate community workflows to avoid libraries' dependencies and maintaining Python modules versions.
These community supported workflows has also few possible configurations, which can be helpful in the future.

I've also added badge to README for both as you suggested.

wiktor2200 commented:
@staticdev I've also started fixing all errors and warnings, you can take a look here: 
https://github.com/wiktor2200/install_nextcloud/tree/fix-ansible-lit-errors-and-warnings
I won't create PR until finished, but feel free to give your feedback and comments :)

staticdev commented:
@wiktor2200 looking great. If you give me permission to our fork, I can fix the rest so you dont have all work.

wiktor2200 commented:
I just pushed second part, Github Actions shows 5 crital issues to go: https://github.com/wiktor2200/install_nextcloud/actions/runs/2640978341

staticdev commented:
@wiktor2200 just a hint.. normally I make sure yamllint is green first before fixing ansible-lint. One practical way to do that is adding to the root `pre-commit-config.yaml`:

```yaml
---
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.1.0
    hooks:
      - id: check-added-large-files
      - id: check-toml
      - id: check-yaml
      - id: end-of-file-fixer
      - id: trailing-whitespace
      - id: check-added-large-files
```

Then by running `pre-commit install && pre-commit run --all-files`. With that, some of the errors are automatically fixed.

wiktor2200 commented:
I guess some changes in configs would be needed for yamllint, but for ansible-lint all test passed: https://github.com/wiktor2200/install_nextcloud/actions/runs/2641022472

staticdev commented:
@wiktor2200 last but not least, you can override line-length and some other configurations with `.yamllint` this is also standard for many projects using yamllint:

https://github.com/staticdev/ansible-role-signal/blob/main/.yamllint

wiktor2200 commented:
That it what I meant by:
> configs would be needed for yamllint
This https://github.com/marketplace/actions/yaml-lint which I used has really simple config add.

wiktor2200 commented:
Both ansible-lint and yamllint succeeded. I've created another PR, please take a look on this one.
#128

staticdev commented on .github/workflows/ansible-lint.yml:
```diff
@@ -0,0 +1,20 @@
+## Source: https://github.com/marketplace/actions/ansible-lint
+
+name: Ansible Lint
+on: [push, pull_request]  # yamllint disable-line rule:truthy
```
To avoid this error Github Actions support an alternative syntax:

```
"on":
  pull_request:
  push:
```

staticdev replied:
Sorry I don't know how to suggest a commit on code review.

staticdev commented on .github/workflows/yamllint.yml:
```diff
@@ -0,0 +1,16 @@
+## Source: https://github.com/marketplace/actions/yaml-lint
+
+name: Yaml Lint
+on: [push, pull_request]  # yamllint disable-line rule:truthy
```
To avoid this error Github Actions support an alternative syntax:

```
"on":
  pull_request:
  push:
```

wiktor2200 replied:
I've already tested that without `"`: https://github.com/aalaesar/install_nextcloud/pull/128/commits/475c40323607e2a008fb5274efed4f4257a2d3f5 and there was still a problem with `truthy` error.

This comment based switch was suggested on official repository of yammlint, there is whole thread about it: https://github.com/adrienverge/yamllint/issues/430 
So actually both solutions are workarounds and we just have to wait for this issue being fixed.

I'd leave it as it is right now, because it's not so obvious that adding `"` would fix/broke whole pipeline, but with this comment it's visible at the first glance :)

staticdev replied:
Ah.. double quotes are needed, I am sure of that. The problem is with on.

aalaesar reviewed (commented):
Great work!
I'll see to enable this on my side.

I feel a bit sad about the state of the CI, so here's some context: 
the role's CI was initially done in travis not for linters but for use case
But more and more cases progressively stopped working in travis and required lots of workarounds to get it working so I eventually gave up 😞.

to guarantee 5 starts with minimal reworks, I suggest to use the [custom yamllint config](https://github.com/ansible/galaxy/blob/devel/galaxy/importer/linters/yamllint.yaml) used by  ansible galaxy as documented [here](https://galaxy.ansible.com/docs/contributing/content_scoring.html).

wiktor2200 commented:
I've updated yamllint configuration. I copied all rules from Ansible Galaxy recommendation and then tweak it a little bit (commented out disabled rules and added them again below with comment and warning severity level).
So all needed checks are done for Galaxy + additional rules will warns about some role's styling issues and has no effect on rating.

wiktor2200 commented:
And when it comes to CI's use cases, Gituhub Actions supports most of use cases. Ansible Molecule tests, deployment to Galaxy or creating a release can be quite easily added even with out-of-the-box modules.

aalaesar commented:
> And when it comes to CI's use cases, Gituhub Actions supports most of use cases. Ansible Molecule tests, deployment to Galaxy or creating a release can be quite easily added even with out-of-the-box modules.

Yes switch test to molecule is probably the good way

LGTM
